// @apibara/protocol/dist/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamDataRequestBuilder = exports.StreamDataRequest = void 0;
exports.StreamDataRequest = {
    /**
     * Start building a `StreamData` request.
     */
    create: () => new StreamDataRequestBuilder(),
};
class StreamDataRequestBuilder {
    request;
    constructor() {
        this.request = {};
    }
    /**
     * Set the new stream id.
     */
    withStreamId(streamId) {
        this.request.streamId = streamId;
        return this;
    }
    /**
     * Set the number of items in each response batch.
     */
    withBatchSize(size) {
        this.request.batchSize = size;
        return this;
    }
    /**
     * Set the cursor from where to resume streaming.
     */
    withStartingCursor(cursor) {
        this.request.startingCursor = cursor;
        return this;
    }
    /**
     * Set the request finality for data.
     */
    withFinality(finality) {
        this.request.finality = finality;
        return this;
    }
    /**
     * Set the stream-specific filter.
     */
    withFilter(filter) {
        this.request.filter = filter;
        return this;
    }
    /**
     * Build and return the request.
     */
    encode() {
        return this.request;
    }
}
exports.StreamDataRequestBuilder = StreamDataRequestBuilder;
